src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp

```diff
@@ -37,29 +37,62 @@
 #include "utilities/align.hpp"
 #include "utilities/debug.hpp"
 
+static int slow_path_size(nmethod* nm) {
+  // The slow path code is out of line with C2
+  return nm->is_compiled_by_c2() ? 0 : 6;
+}
+
+// This is the offset of the entry barrier from where the frame is completed.
+// If any code changes between the end of the verified entry where the entry
+// barrier resides, and the completion of the frame, then
+// NativeNMethodCmpBarrier::verify() will immediately complain when it does
+// not find the expected native instruction at this offset, which needs updating.
+// Note that this offset is invariant of PreserveFramePointer.
+static int entry_barrier_offset(nmethod* nm) {
+  BarrierSetAssembler* bs_asm = BarrierSet::barrier_set()->barrier_set_assembler();
+  switch (bs_asm->nmethod_patching_type()) {
+  case NMethodPatchingType::stw_instruction_and_data_patch:
+    return -4 * (4 + slow_path_size(nm));
+  case NMethodPatchingType::conc_instruction_and_data_patch:
+    return -4 * (10 + slow_path_size(nm));
+  case NMethodPatchingType::conc_data_patch:
+    return -4 * (5 + slow_path_size(nm));
+  }
+  ShouldNotReachHere();
+  return 0;
+}
+
 class NativeNMethodBarrier: public NativeInstruction {
   address instruction_address() const { return addr_at(0); }
 
-  int guard_offset() {
-    BarrierSetAssembler* bs_asm = BarrierSet::barrier_set()->barrier_set_assembler();
-    if (bs_asm->nmethod_code_patching()) {
-      return 4 * 15;
-    } else {
-      return 4 * 10;
-    }
+  int local_guard_offset(nmethod* nm) {
+    // It's the last instruction
+    return (-entry_barrier_offset(nm)) - 4;
   }
 
-  int *guard_addr() {
-    return reinterpret_cast<int*>(instruction_address() + guard_offset());
+  int *guard_addr(nmethod* nm) {
+    if (nm->is_compiled_by_c2()) {
+      // With c2 compiled code, the guard is out-of-line in a stub
+      // We find it using the RelocIterator.
+      RelocIterator iter(nm);
+      while (iter.next()) {
+        if (iter.type() == relocInfo::entry_guard_type) {
+          entry_guard_Relocation* const reloc = iter.entry_guard_reloc();
+          return reinterpret_cast<int*>(reloc->addr());
+        }
+      }
+      ShouldNotReachHere();
+    }
```

Comment:
I'm guessing this function is not performance-critical.  If it was, then we could consider adding a new field to the nmethod to keep track of the offset for the guard.

Reply:
Yeah I thought about adding a new entry to CodeOffsets, but it seemed like this C2-only AArch64-only detail seemed a bit noisy in the shared code then. And it didn't seem warranted, given how cold this path is by design.